### PR TITLE
Add gpu resource for remote-run and cli (for hook validation)

### PR DESCRIPTION
**Pull request by charleskwwan**
## Summary
- add gpus resource to remote-run and basic instance config (in utils)
- add gpus functions for basic instance config
- add gpus to adhoc service schema (used by py-gitolite hooks to check yelpsoa-configs)
- add tests for gpus in utils; none for remote-run since no tests yet
- add gpus to paasta metastatus as another resource
- modify the ResourceInfo namedtuple to have gpus default to 0
## Testing
- tox
- manual testing for instance gpus
- manual testing for adhoc service schema - added a gpus instance in yelpsoa-config's taskproc-canary service and ran it against my schema using cli.cmds.validate
## Notes
- I will run an end-to-end test once I have updated yelpsoa-configs - validation hooks should use paasta's updated gpus schema
- Does anyone know if I need to update any version numbers or what the deployment path is to guarantee the hooks will use my changes?
- Not sure if metastatus for gpus is strictly necessary since we don't care about its threshold and we never reserve any

### CODE REVIEW

File: paasta_tools/paasta_remote_run.py [hunk 1]
```diff
@@ -171,6 +172,8 @@ def paasta_to_task_config_kwargs(
         'uris': [uris],
         'docker_parameters': docker_parameters,
     }
+    if gpus > 0:
+        kwargs['gpus'] = float(gpus)
```
Comment (solarkennedy) [inline paasta_tools/paasta_remote_run.py]: is it really a float?

File: paasta_tools/utils.py [hunk 2]
```diff
@@ -410,6 +419,12 @@ def check_disk(self):
                 return False, 'The specified disk value "%s" is not a valid float or int.' % disk
         return True, ''
 
+    def check_gpus(self):
```
Comment (solarkennedy) [inline paasta_tools/utils.py]: I think this is overkill, can you instead add it to the json schema?

Comment (solarkennedy) [inline paasta_tools/utils.py]: Wait sorry, you did. :)

**Comment by charleskwwan** [comment]
@nhandler told me that paasta metastatus also needs to be updated to include gpus, so I've pushed another change for that.

File: paasta_tools/paasta_remote_run.py (hunk 1 above)
Comment (charleskwwan) [inline paasta_tools/paasta_remote_run.py]: doing this because task processing and mesos expects floats, even though gpus cannot be partially used.

File: paasta_tools/cli/schemas/adhoc_schema.json [hunk 3]
```diff
@@ -26,6 +26,12 @@
                 "exclusiveMinimum": true,
                 "default": 1024
             },
+            "gpus": {
```
Comment (sagar8192) [inline paasta_tools/cli/schemas/adhoc_schema.json]: If gpus cannot be used partially, we probably want to enforce that here. This should be an int instead? Also, if we don't want to allow fractions for gpus, we should change taskproc to only accept ints.

File: paasta_tools/metrics/metastatus_lib.py [hunk 4]
```diff
@@ -32,9 +32,10 @@
 from paasta_tools.utils import PaastaColors
 from paasta_tools.utils import print_with_indent
 
-
+class ResourceInfo(namedtuple('ResourceInfo', ['cpus', 'mem', 'disk', 'gpus'])):
```
Comment (sagar8192) [inline paasta_tools/metrics/metastatus_lib.py]: Instead of inheriting from namedtuple, can we just make cpus, mem, disk and gpus class members?

File: paasta_tools/metrics/metastatus_lib.py [hunk 5]
```diff
@@ -78,8 +79,20 @@ def get_mesos_disk_status(metrics):
     return total, used, available
 
 
+def get_mesos_gpu_status(metrics):
+    """Takes in the mesos metrics and analyzes them, returning gpus status.
+
+    :param metrics: mesos metrics dictionary
+    :returns: Tuple of the output array and is_ok bool
+    """
+    total = metrics['master/gpus_total']
```
Comment (sagar8192) [inline paasta_tools/metrics/metastatus_lib.py]: Are these available in master right now? I think mesos only adds it if we have slaves with gpus in the cluster. It won't have it otherwise.  Let's check if this keys is available or not.

File: paasta_tools/metrics/metastatus_lib.py (hunk 4 above)
Comment (charleskwwan) [inline paasta_tools/metrics/metastatus_lib.py]: I can't because there are parts of paasta that used namedtuple specific functions.

File: paasta_tools/metrics/metastatus_lib.py (hunk 5 above)
Comment (charleskwwan) [inline paasta_tools/metrics/metastatus_lib.py]: There are available as far as know. I checked mesosstage, which has never had GPUs and it has them: http://mesos.paasta-mesosstage.yelp:5050/metrics/snapshot

File: paasta_tools/cli/schemas/adhoc_schema.json (hunk 3 above)
Comment (charleskwwan) [inline paasta_tools/cli/schemas/adhoc_schema.json]: Made change to taskproc, will follow through here.

File: paasta_tools/autoscaling/autoscaling_cluster_lib.py [hunk 6]
```diff
@@ -213,10 +213,13 @@ def get_mesos_utilization_error(
...
         total_pool_resources = region_pool_utilization_dict['total']
-        utilization = 1.0 - min([
-            float(float(pair[0]) / float(pair[1]))
-            for pair in zip(free_pool_resources, total_pool_resources)
-        ])
+        used_percs = []
+        for pair in zip(free_pool_resources, total_pool_resources):
+            try:
+                used_percs.append(float(float(pair[0]) / float(pair[1])))
+            except ZeroDivisionError:  # should occur for 0 gpus, which is valid
```
Comment (keymone) [inline paasta_tools/autoscaling/autoscaling_cluster_lib.py]: if you can easily check for error condition there's no reason to use exceptions:

    for used, total in zip(free_pool_resources, total_pool_resources):
        if math.isclose(total, 0):
            continue
        used_percs.append(float(used)/float(total))

Comment (charleskwwan) [inline paasta_tools/autoscaling/autoscaling_cluster_lib.py]: Good point! I'll make the change.
